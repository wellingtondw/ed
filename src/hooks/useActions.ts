import { useDispatch } from 'react-redux';
import { bindActionCreators } from 'redux';

import {
  popularMoviesRequest,
  searchMoviesRequest,
  resetMoviesSearch
} from '../store/modules/movies/actions';

export const useActions = () => {
  const dispatch = useDispatch();

  return bindActionCreators(
    {
      popularMoviesRequest,
      searchMoviesRequest,
      resetMoviesSearch
    },
    dispatch
  );
};
