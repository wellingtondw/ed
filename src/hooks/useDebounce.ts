import { useRef } from 'react';

export const useDebounce = (fn: any, delay: number) => {
  const timeoutRef: any = useRef(null);

  function debouncedFn(...args: any) {
    window.clearTimeout(timeoutRef.current);

    timeoutRef.current = window.setTimeout(() => {
      fn(...args);
    }, delay);
  }

  return debouncedFn;
};
