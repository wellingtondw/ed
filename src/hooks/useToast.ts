import { toast } from 'react-toastify';

type ToastProps = {
  message: string;
  type?: 'error' | 'success';
};

export function useToast({ message, type }: ToastProps) {
  switch (type) {
    case 'success':
      return toast.error(message);
    case 'error':
    default:
      return toast.error(message);
  }
}
