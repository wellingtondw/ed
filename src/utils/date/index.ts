type GetFullYearParams = {
  date: string | Date;
};
export const getFullYear = ({ date }: GetFullYearParams) => {
  return new Date(date).getFullYear();
};
