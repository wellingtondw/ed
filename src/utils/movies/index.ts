import { MovieCardListDTO } from 'components/MovieCardList';

export type MoviesDTO = {
  poster_path: string;
  title: string;
  release_date: string;
  vote_average: number;
  id: number;
};

type FormatMoviesListParams = {
  items: MoviesDTO[];
};

export const formatMoviesList = ({ items }: FormatMoviesListParams) => {
  return items.map((item): MovieCardListDTO => {
    const { poster_path, title, release_date, vote_average, id } = item;

    return {
      posterImage: poster_path,
      title,
      date: release_date,
      rating: vote_average,
      id
    };
  });
};
