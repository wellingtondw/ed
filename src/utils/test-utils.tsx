import { createStore } from 'redux';
import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import { createMemoryHistory } from 'history';

import { ThemeProvider } from 'styled-components';
import theme from 'styles/theme';
import { Router } from 'react-router-dom';
import { Provider } from 'react-redux';

import rootReducer from '../store/modules/rootReducer';

const memoryHistory = createMemoryHistory();

const customRender = (
  ui: ReactElement,
  initialState = {},
  store = createStore(rootReducer, initialState),
  history = memoryHistory,
  { ...renderOptions }: RenderOptions = {}
) =>
  render(
    <Provider store={store}>
      <Router navigator={history} location={history.location}>
        <ThemeProvider theme={theme}>{ui}</ThemeProvider>
      </Router>
    </Provider>,
    renderOptions
  );

export * from '@testing-library/react';
export { customRender as render };
