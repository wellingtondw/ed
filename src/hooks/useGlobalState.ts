import { useSelector } from 'react-redux';
import { IMoviesState } from '../store/modules/movies/types';
import { IMovieState } from '../store/modules/movie/types';

export interface IState {
  movies: IMoviesState;
  movie: IMovieState;
}

export const useGlobalState = (): IState => {
  return useSelector<IState>((state) => state) as IState;
};
