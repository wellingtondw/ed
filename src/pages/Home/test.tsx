import { render, fireEvent, screen, waitFor } from '../../utils/test-utils';
import * as actions from '../../hooks/useActions';

import { Home } from '.';

const popularMoviesRequest = jest.fn();
jest.spyOn(actions, 'useActions').mockReturnValue({ popularMoviesRequest } as any);

jest.mock('../../components/Header', () => {
  return {
    __esModule: true,
    Header: function Mock() {
      return <div data-testid="Mock Header"></div>;
    }
  };
});

const mockSetSearchParams = jest.fn();
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useSearchParams: () => [{ get: jest.fn() }, mockSetSearchParams]
}));

const mockToast = jest.fn();
jest.mock('../../hooks/useToast', () => {
  return {
    useToast: () => mockToast()
  };
});

const initialStateMock = {
  movies: {
    popularMovies: { data: [], loading: false, totalCount: 200 }
  }
};

const sut = (initialState = initialStateMock) => {
  return render(<Home />, initialState);
};

describe('<Home />', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should be able to render correctly', () => {
    const { container } = sut();

    expect(container).toMatchSnapshot();
  });

  it('should be able to load popular movies on render Home page', () => {
    sut();

    expect(popularMoviesRequest).toHaveBeenCalledTimes(1);
  });

  it('should be able to load popular movies with pagination', async () => {
    sut();

    const page = '3';
    const paginationButton = await waitFor(() => screen.getByRole('button', { name: page }));
    fireEvent.click(paginationButton);

    expect(popularMoviesRequest).toBeCalledTimes(2);
    expect(popularMoviesRequest).toBeCalledWith(Number(page));
  });

  it('should not be able to render Pagination component if loading', () => {
    sut({
      movies: {
        ...initialStateMock.movies,
        popularMovies: {
          ...initialStateMock.movies.popularMovies,
          loading: true
        }
      }
    });

    expect(screen.queryByRole('button', { name: '1' })).not.toBeInTheDocument();
  });

  it('should be able to handle fetch popular movies error', () => {
    const currentPage = 10;
    sut({
      movies: {
        ...initialStateMock.movies,
        popularMovies: {
          ...initialStateMock.movies.popularMovies,
          error: true,
          currentPage
        }
      } as any
    });

    expect(mockToast).toHaveBeenCalledTimes(1);
    expect(mockSetSearchParams).toHaveBeenCalledWith({ page: String(currentPage) });
  });
});
