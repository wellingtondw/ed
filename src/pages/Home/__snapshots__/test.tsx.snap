// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Home /> should be able to render correctly 1`] = `
<div>
  <div
    data-testid="Mock Header"
  />
  <div
    class="sc-kDDrLX bjkhes"
  >
    <ul
      class="sc-papXJ gxDyOm"
    />
    <div
      class="sc-iqcoie fbUbgs"
    >
      <div
        class="sc-crXcEl iKqUVm"
      >
        <button
          class="sc-breuTD CmIAE"
        >
          1
        </button>
        <button
          class="sc-breuTD fEpflM"
        >
          2
        </button>
        <button
          class="sc-breuTD fEpflM"
        >
          3
        </button>
        <span
          class="sc-evZas gSzlv"
        >
          . . .
        </span>
        <button
          class="sc-breuTD fEpflM"
        >
          Infinity
        </button>
      </div>
    </div>
  </div>
</div>
`;
